Pers/vprokhorenko/homework 14

Traknar2142 opened the pull request:
Реализация дз №14

Traknar2142 committed:
Реализация ДЗ №14

Traknar2142 committed:
Реализация ДЗ №14

stvort commented on homework-14/src/main/java/ru/otus/homework/service/BookConvertationService.java:
```diff
@@ -0,0 +1,39 @@
+        author.put("_id", new ObjectId());
+        author.put("name", book.getAuthor().getName());
+
+        mongoTemplate.insert(author, "author").getObjectId("_id");
```
Это обязанность врайтера. Он это делает пакетно. В этом основная фишка Spring**Batch**, что видно из названия. А процессор вызывается для каждого айтема

Traknar2142 committed:
Правки ДЗ №14

Traknar2142 committed:
Разбил добавление авторов, жанров, книг на шаги
Перенес манипуляции с mongoTemplate из процессора во врайтер

stvort commented on homework-14/src/main/java/ru/otus/homework/writer/BookMongoItemWriter.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Override
+    public void write(List<? extends Document> documents) {
+        for (Document document : documents) {
+            Document author = (Document) document.get("author");
+
+            Document genre = (Document) document.get("genre");
+
+            Query authorQuery = new Query();
+            authorQuery.addCriteria(where("name").is(author.getString("name")));
+
+            Query genreQuery = new Query();
+            genreQuery.addCriteria(where("name").is(genre.getString("name")));
+
+            MAuthor mAuthor = mongoTemplate.findOne(authorQuery, MAuthor.class);
+
+            MGenre mGenre = mongoTemplate.findOne(genreQuery, MGenre.class);
+
+            document.put("author", mAuthor);
+            document.put("genre", mGenre);
+            mongoTemplate.insert(document, "book");
+        }
+    }
```
Теперь мы написали свой врайтер, который пишет айтемы по одному. Чтобы операция записи была пакетной, не достаточно просто реализовать интерфейс, надо сделать в реализации пакетную запись. Например, сделав наследника `MongoItemWriter` и/или изучив его код

stvort commented on homework-14/src/main/java/ru/otus/homework/writer/BookMongoItemWriter.java:
```diff
@@ -0,0 +1,44 @@
+            Query genreQuery = new Query();
+            genreQuery.addCriteria(where("name").is(genre.getString("name")));
+
+            MAuthor mAuthor = mongoTemplate.findOne(authorQuery, MAuthor.class);
+
+            MGenre mGenre = mongoTemplate.findOne(genreQuery, MGenre.class);
```
Ну и мы можем не лазить в БД за авторами и жанрами. Они у нас пишутся в приложении, у нас есть их идентификаторы. Мы их можем запомнить после записи или даже сгенерировать до нее (`new ObjectId`)

stvort commented on homework-14/src/main/java/ru/otus/homework/writer/BookMongoItemWriter.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Override
+    public void write(List<? extends Document> documents) {
```
Кстати с `Document` мы тоже не обязаны работать, можно с книгами/авторами/жанрами

Traknar2142 committed:
- Удалил самописные врайтеры
- Добавил кеширование ID для дальнейшего мапинга их в сущности
- Изменил в процессоре книг работу с Document на работу с MBook

Traknar2142 committed:
Убрал лишнее

Traknar2142 committed:
фикс бага с заканчивающимся id

Traknar2142 committed:
Merge branch 'master' into pers/vprokhorenko/homework-14
